pkg/planner/core/casetest/mpp/mpp_test.go

```diff
@@ -865,3 +865,106 @@ func TestIssue52828(t *testing.T) {
 	tk.MustQuery("explain SELECT MAX( OUTR . col_int ) AS X FROM C AS OUTR2 INNER JOIN B AS OUTR ON ( OUTR2 . col_decimal_not_null = OUTR . col_decimal_not_null AND OUTR2 . pk = OUTR . col_int_not_null ) " +
 		"WHERE OUTR . col_decimal_not_null IN ( SELECT INNR . col_int_not_null + 1 AS Y FROM DD AS INNR WHERE INNR . pk > INNR . pk OR INNR . col_varchar_10_not_null >= INNR . col_varchar_10 ) GROUP BY OUTR . col_datetime_not_null")
 }
+
+func TestMPPJoinWithRemoveUselessExchange(t *testing.T) {
+	store, dom := testkit.CreateMockStoreAndDomain(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+	tk.MustExec(`CREATE TABLE t1 (
+    v1 INT NOT NULL,
+    v2 INT NOT NULL,
+    PRIMARY KEY (v1)
+);`)
+	tk.MustExec(`CREATE TABLE t2 (
+    v1 INT NOT NULL,
+    v2 INT NOT NULL,
+    PRIMARY KEY (v1)
+);`)
+	tk.MustExec(`CREATE TABLE t3 (
+    v1 INT NOT NULL,
+    v2 INT NOT NULL,
+    PRIMARY KEY (v1)
+);`)
+	tk.MustExec(`CREATE TABLE t4 (
+    v1 INT NOT NULL,
+    v2 INT NOT NULL,
+    PRIMARY KEY (v1)
+);`)
+	testkit.SetTiFlashReplica(t, dom, "test", "t1")
+	testkit.SetTiFlashReplica(t, dom, "test", "t2")
+	testkit.SetTiFlashReplica(t, dom, "test", "t3")
+	testkit.SetTiFlashReplica(t, dom, "test", "t4")
+	tk.MustQuery(`explain format = 'brief' SELECT /*+ SHUFFLE_JOIN(aaa, bbb) */ count(*)
```

Comment:
Better to put those query and result into  xxx_in.json and xxx_out.json

Reply:
it has rewritten to this type.